File: C++/025. Reverse Nodes in k-Group.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *res = new ListNode(0), *first = res;
        first->next = head;
        while (true) {
            ListNode *last = first;
            for (int i = 0; i < k && last != nullptr; i++)
                last = last->next;
            if (last == nullptr)
                break;
            last = first->next;
            for (int i = 1; i < k; i++) {
                ListNode *tmp = last->next;
                last->next = last->next->next;
                tmp->next = first->next;
                first->next = tmp;
            }
            first = last;
        }
        return res->next;
    }
};

File: Go/025. Reverse Nodes in k-Group.go
package main

type ListNode struct {
    Val int
    Next *ListNode
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	res := &ListNode{Next:head}
	first := res
	for true {
		i, last := 0, first
		for ; i < k && last != nil; i++ {
			last = last.Next
		}
		if last == nil {
			break
		}
		last = first.Next
		for i = 1; i < k; i++ {
			tmp := last.Next
			last.Next = last.Next.Next
			tmp.Next = first.Next
			first.Next = tmp
		}
		first = last
	}
	return res.Next
}